main.py

```diff
@@ -670,6 +670,31 @@ def delete_link_metadata(self, request: Request) -> JSONResponse:
         self.notify_topology_update()
         return JSONResponse("Operation successful")
 
+    @rest('v3/links/{link_id}', methods=['DELETE'])
+    def delete_link(self, request: Request) -> JSONResponse:
+        """Delete a down link from topology. It won't work for up links."""
+        link_id = request.path_params["link_id"]
+        try:
+            link = self.links[link_id]
+        except KeyError:
+            raise HTTPException(404, detail="Link not found.")
+        if link.status != EntityStatus.DOWN:
+            raise HTTPException(409, detail="Link is not down.")
+        if link.endpoint_a.link and link == link.endpoint_a.link:
+            switch = link.endpoint_a.switch
+            link.endpoint_a.link = None
+            link.endpoint_a.nni = False
+            self.topo_controller.upsert_switch(switch.id, switch.as_dict())
+        if link.endpoint_a.link and link == link.endpoint_b.link:
+            switch = link.endpoint_b.switch
+            link.endpoint_b.link = None
+            link.endpoint_b.nni = False
+            self.topo_controller.upsert_switch(switch.id, switch.as_dict())
+        link = self.links.pop(link_id)
```

Comment:
On this like you have a potential race condition since `self.links` is being r/w accessed by multiple threads, so multiple threads could get suspended and resume at this point here. I'd recommend acquiring `self._link_locks` at this point here while trying to pop the link_id, typically we also try to avoid holding locks for too long specially in an endpoint that might end up doing too much IO (misbehaving client, until we have request rate limiting in place), for instance, something like:

```
        try:
          with self._links_lock:
            link = self.links.pop(link_id)
          self.topo_controller.delete_link(link_id)
          self.notify_topology_update()
          return JSONResponse("Operation successful")
        except KeyError:
          raise HTTPException(404, detail="Link not found")

Reply:
Using `self._links_lock` but database updates are outside.